File: leetcode/solutions/1005.MaximizeSumOfArrayAfterKNegations/largestSumAfterKNegations.go
package solution

import (
	"sort"
)

func largestSumAfterKNegations(A []int, K int) int {
	sum := 0
	sort.Ints(A)
	var minAbs int

	if A[0] < 0 {
		minAbs = -A[0]
	} else {
		minAbs = A[0]
	}

	for _, num := range A {
		if num <= 0 && K > 0 {
			num = -num
			K--

			if minAbs > num {
				minAbs = num
			}
		}

		sum += num
	}

	if K%2 == 1 {
		sum -= minAbs * 2
	}

	return sum
}

File: leetcode/solutions/1005.MaximizeSumOfArrayAfterKNegations/largestSumAfterKNegations.ts
export default function largestSumAfterKNegations(
  A: number[],
  K: number
): number {
  A.sort((a, b) => a - b);

  let sum = 0;
  let min = Infinity;
  let zeroExists = false;

  for (let [i, num] of A.entries()) {
    if (num < 0 && K > 0) {
      num = -num;
      K--;
    }
    if (num === 0) {
      zeroExists = true;
    }

    sum += num;

    if (!zeroExists) {
      min = Math.min(min, num);
    }
  }

  if (!zeroExists && K % 2 !== 0) {
    sum -= min * 2;
  }

  return sum;
}
